Switch from sleep/wakeup to unavailable/available callbacks

jcredding opened the pull request:
This switches workers from having sleep and wakeup callbacks to
having unavailable and available callbacks. This also switches
from counting the number of waiting workers to a set of available
workers. The goal of this is to be more accurate about the workers
behavior and make it easier to understand when callbacks will
be called.

The workers previous sleep and wakeup callbacks were not always
accurate. It's not possible for the worker to tell if it will
actually go to sleep and wakeup. This switches out the inaccuracy
of those callbacks for worker unavailable/available logic. This
happens when it starts working (it becomes unavailable) and
once it finished working (it becomes available). This is accurate
and less confusing than the previous sleep and wakeup callbacks.

As part of adding in the unavailable/available callbacks, the
`work` method has been made private and renamed to `dwp_work`. This
method was originally public because it was intended to be used as
a test helper. With the test runner though, there is a better
interface to use. This eases the maintenance of the worker because
now we don't also need to worry about keeping its work method
functioning as a test helper.

This also switches the workers waiting count on the runner to an
available workers set. Similarly to the callbacks, this is more
accurate than the workers waiting count. Instead of keeping up with
a count, we add and remove worker object ids from a set. This can
then be used to query whether or not a worker is available or not.

Finally, this removes the `num_workers` and changes the `waiting`
methods from the dat-worker-pool. The `num_workers` method is
confusing because its how many were configured not how many are
actually running. The `waiting` method is now an
`available_worker_count` which returns the number of workers that
are available to do work.

@kellyredding - Ready for review.

kellyredding commented:
@jcredding "... to a set of available *worekrs*." in first paragraph of commit msg.

kellyredding commented on lib/dat-worker-pool/worker.rb:
```diff
@@ -50,9 +50,16 @@ def initialize(runner, queue)
-        work!(work_item)
-        dwp_run_callback('after_work', work_item)
+        @dwp_runner.mark_worker_unavailable(self)
+        dwp_run_callback 'on_unavailable'
```
@jcredding I think you need to move this run callback line into the begin..rescue so that if an error occurs the work gets marked available, right?

kellyredding commented on lib/dat-worker-pool/worker.rb:
```diff
@@ -187,8 +189,6 @@ def initialize(worker_class, options = nil)
-          self.wakeup
           self.work(work_item)
           self.shutdown
         end
```
@jcredding it sucks that we don't have more granularity on this.  The work method now implies 5 hooks for custom logic and you can't run them one at a time.  I'd love it if we could have it go: `make_unavailable`, `work`, `make_available`.  Is that possible?

kellyredding commented on lib/dat-worker-pool.rb:
```diff
@@ -51,16 +51,8 @@ def add_work(work_item)
-
-  def waiting
-    @runner.workers_waiting_count
-  end
```
@jcredding it would be great to get an `available_worker_count` or something like that method on both the worker pool and the runner.  I'd think it could have utility and seems pretty cheap to offer, right?

kellyredding commented on test/system/use_worker_pool_tests.rb:
```diff
@@ -61,29 +61,26 @@ def work!(seconds); sleep(seconds); end
-      # only one worker should be waiting
+      # only one worker should be available
       subject.add_work 5
-      assert_equal 1,    subject.waiting
```
@jcredding case-in-point on the "available worker count" method: I don't love that you state only one worker should be available but don't prove it with an assertion.

kellyredding commented:
@jcredding a few comments.  I'd love your thoughts on them.

jcredding commented:
@kellyredding - I've updated this with your feedback. I also went ahead and made the `work` method private (and renamed it to `dwp_work`) like we discussed. Thanks for the suggestions.

kellyredding commented:
@jcredding yeah, I'm cool with all this - thanks for the updates! :-1:
